File: apps/ruby/design-pattern/07-composite/lib/composite_task.rb
# frozen_string_literal: true

require_relative "task"

# 複合タスククラス（Composite パターン - Composite）
class CompositeTask < Task
  def initialize(name)
    super(name)
    @sub_tasks = []
  end

  def add_sub_task(task)
    @sub_tasks << task
    task.parent = self
  end

  def remove_sub_task(task)
    @sub_tasks.delete(task)
    task.parent = nil
  end

  def [](index)
    @sub_tasks[index]
  end

  def []=(index, task)
    @sub_tasks[index] = task
    task.parent = self
  end

  def get_time_required
    @sub_tasks.sum(&:get_time_required)
  end

  def total_number_basic_tasks
    @sub_tasks.sum(&:total_number_basic_tasks)
  end
end

File: apps/java/design-pattern/src/main/java/pattern/composite/CompositeTask.java
package pattern.composite;

import java.util.ArrayList;
import java.util.List;

/**
 * 複合タスククラス（Composite パターン - Composite）
 */
public class CompositeTask extends Task {

    private final List<Task> subTasks = new ArrayList<>();

    public CompositeTask(String name) {
        super(name);
    }

    public void addSubTask(Task task) {
        subTasks.add(task);
        task.setParent(this);
    }

    public void removeSubTask(Task task) {
        subTasks.remove(task);
        task.setParent(null);
    }

    public Task getSubTask(int index) {
        return subTasks.get(index);
    }

    @Override
    public double getTimeRequired() {
        return subTasks.stream()
                .mapToDouble(Task::getTimeRequired)
                .sum();
    }

    @Override
    public int getTotalBasicTasks() {
        return subTasks.stream()
                .mapToInt(Task::getTotalBasicTasks)
                .sum();
    }
}
